File: 2109_-_Adding_Spaces_To_String/2109.cpp
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    string addSpaces(string s, vector<int> &spaces) {
        int resultIndex = 0;
        int spacesIndex = 0;
        const int spacesSize = spaces.size();
        string result(s.size() + spaces.size(), ' ');

        for (int stringIndex = 0; stringIndex < s.size(); stringIndex++) {
            if (spacesIndex < spacesSize &&
                spaces[spacesIndex] == stringIndex) {
                ++spacesIndex;
                ++resultIndex;
            }
            result[resultIndex] = s[stringIndex];
            resultIndex++;
        }

    return result;
    }
};

File: 2109_-_Adding_Spaces_To_String/2109.py
class Solution:
    def addSpaces(self, s: str, spaces: list[int]) -> str:
        """
        Turn each word into it's own string and add to a list.
        Join the list with a space.

        Much MUCH faster than concatenating strings.
        """
        string_as_list = []

        string_as_list.append(s[:spaces[0]])

        for i in range(len(spaces) - 1):
            string_as_list.append(s[spaces[i]:spaces[i + 1]])

        string_as_list.append(s[spaces[-1]:])

        return " ".join(string_as_list)
